drivers/gpio/gpio_numaker.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_gpio
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <soc.h>
+
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#endif
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include "NuMicro.h"
+
+#include <zephyr/logging/log.h>
+
+#define NU_MFP_POS(pinindex) ((pinindex % 4) * 8)
+
+LOG_MODULE_REGISTER(gpio_numaker, LOG_LEVEL_ERR);
+
+struct gpio_numaker_config {
+	struct gpio_driver_config common;
+	uint32_t reg;
+	uint32_t gpa_base;
+	uint32_t size;
+	uint32_t clk_modidx;
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+	const struct device *clk_dev;
+#endif
+};
+
+struct gpio_numaker_data {
+	struct gpio_driver_data common;
+	sys_slist_t callbacks;
+};
+
+static int gpio_numaker_configure(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct gpio_numaker_config *config = dev->config;
+	struct gpio_numaker_data *data = dev->data;
+	GPIO_T *gpio_base = (GPIO_T *)config->reg;
+	uint32_t pinMfpMask = (0x1f << NU_MFP_POS(pin));
+	uint32_t pinMask = BIT(pin); /* mask for pin index --> (0x01 << pin) */
+	int err = 0;
+
+	ARG_UNUSED(data);
+
+	/* Check for an invalid pin number */
+	if (pin >= 15) {
+		return -EINVAL;
+	}
+
+	SYS_UnlockReg();
+
+	/* Enable GPIO clock */
+#ifdef CONFIG_CLOCK_CONTROL_NUMAKER_SCC
+	struct numaker_scc_subsys scc_subsys;
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = config->clk_modidx;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	err = clock_control_on(config->clk_dev, (clock_control_subsys_t)&scc_subsys);
+	if (err != 0) {
+		goto move_exit;
+	}
+#else
+	/* Enable GPIO module clock */
+	CLK_EnableModuleClock(config->clk_modidx);
+#endif
+
+	/* Configure GPIO direction */
+	switch (flags & GPIO_DIR_MASK) {
+	case GPIO_INPUT:
+		GPIO_SetMode(gpio_base, pinMask, GPIO_MODE_INPUT);
+		break;
+	case GPIO_OUTPUT:
+		GPIO_SetMode(gpio_base, pinMask, GPIO_MODE_OUTPUT);
+		break;
+	case (GPIO_INPUT | GPIO_OUTPUT):
+		GPIO_SetMode(gpio_base, pinMask, GPIO_MODE_QUASI);
+		break;
+	default:
+		err = -ENOTSUP;
+		goto move_exit;
+	}
+
+	if (flags & GPIO_LINE_OPEN_DRAIN) {
+		GPIO_SetMode(gpio_base, pinMask, GPIO_MODE_OPEN_DRAIN);
+	}
+
+	/* Set Multi-function, default is GPIO */
+	uint32_t port_index =
+	(config->reg - config->gpa_base) / config->size;
```

Comment:
odd indentation: bring on the previous line

Reply:
Previous version with macro of get-base & get-size, so got WARNING:LONG_LINE at that time.
Now, it could be under limit of long line. 
Thanks of your remind.